File: src/NodeParser/HiddenTypeNodeParser.ts
import type ts from "typescript";
import type { Context } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { HiddenType } from "../Type/HiddenType.js";
import { isNodeHidden } from "../Utils/isHidden.js";

export class HiddenNodeParser implements SubNodeParser {
    public constructor(protected typeChecker: ts.TypeChecker) {}

    public supportsNode(node: ts.KeywordTypeNode): boolean {
        return isNodeHidden(node);
    }

    public createType(_node: ts.KeywordTypeNode, _context: Context): BaseType {
        return new HiddenType();
    }
}

File: internal/parser/hidden_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/tsutils"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// HiddenNodeParser turns nodes annotated with the @hidden JSDoc tag into
// HiddenType (src/NodeParser/HiddenTypeNodeParser.ts).
type HiddenNodeParser struct {
	typeChecker *checker.Checker
}

func NewHiddenNodeParser(typeChecker *checker.Checker) *HiddenNodeParser {
	return &HiddenNodeParser{typeChecker: typeChecker}
}

func (p *HiddenNodeParser) SupportsNode(node *ast.Node) bool {
	return tsutils.IsNodeHidden(node)
}

func (p *HiddenNodeParser) CreateType(_ *ast.Node, _ *Context, _ *types.ReferenceType) types.Type {
	return &types.HiddenType{}
}
